File: backend/app/services/booking.py
from datetime import datetime, timedelta

from fastapi import HTTPException, status
from sqlalchemy import and_, func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.config import settings
from app.models import (
    Availability,
    Booking,
    BookingSource,
    BookingStatus,
    Guest,
    GuestVerificationStatus,
    Payment,
    PaymentGateway,
    PaymentStatus,
    Property,
    RatePlan,
    RoomType,
)
from app.schemas.admin_content import AdminBookingCreate
from app.schemas.public import BookingCreate, GuestIn

# Statuses that count as "holding" a room for a given night.
HOLDING_STATUSES = (BookingStatus.confirmed, BookingStatus.pending_payment)
# A guest can cancel a booking themselves as long as it's still in one of these
# states and their stay hasn't already started.
CANCELLABLE_STATUSES = (BookingStatus.pending_payment, BookingStatus.confirmed, BookingStatus.pending_whatsapp)
MAX_STAY_NIGHTS = 30
MAX_ADVANCE_BOOKING_DAYS = 730  # 2 years — keeps the availability loop bounded


def _nights(check_in, check_out) -> list:
    return [check_in + timedelta(days=i) for i in range((check_out - check_in).days)]


async def _held_count(session: AsyncSession, room_type_id: int, day) -> int:
    expiry_cutoff = datetime.utcnow() - timedelta(minutes=settings.pending_payment_expiry_minutes)
    result = await session.execute(
        select(func.count(Booking.id)).where(
            Booking.room_type_id == room_type_id,
            Booking.check_in <= day,
            Booking.check_out > day,
            and_(
                Booking.status.in_(HOLDING_STATUSES),
                # an expired pending_payment hold no longer counts
                ~and_(
                    Booking.status == BookingStatus.pending_payment,
                    Booking.created_at < expiry_cutoff,
                ),
            ),
        )
    )
    return result.scalar_one()


async def _validate_and_lock(
    session: AsyncSession,
    *,
    property_id: int,
    room_type_id: int,
    rate_plan_id: int,
    check_in,
    check_out,
    guests_count: int,
    enforce_capacity: bool,
) -> tuple[Property, RoomType, RatePlan, list]:
    if check_in >= check_out:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "check_out must be after check_in")
    if check_in < datetime.utcnow().date():
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "check_in cannot be in the past")
    if (check_out - check_in).days > MAX_STAY_NIGHTS:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, f"Stays longer than {MAX_STAY_NIGHTS} nights aren't supported")
    if (check_in - datetime.utcnow().date()).days > MAX_ADVANCE_BOOKING_DAYS:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "Check-in date is too far in the future")

    room_type = await session.get(RoomType, room_type_id)
    rate_plan = await session.get(RatePlan, rate_plan_id)
    property_ = await session.get(Property, property_id)
    if not room_type or not rate_plan or not property_:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Property, room type, or rate plan not found")
    if rate_plan.room_type_id != room_type.id or room_type.property_id != property_.id:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "Room/rate plan does not belong to this property")

    if guests_count > room_type.max_occupancy:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "Guest count exceeds room's max occupancy")
    if not (rate_plan.valid_from <= check_in and check_out <= rate_plan.valid_to):
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "Dates fall outside this rate plan's validity window")

    nights = _nights(check_in, check_out)

    # Lock the relevant availability rows so concurrent bookings for the same
    # room/date range serialize instead of both reading a stale remaining count.
    lock_result = await session.execute(
        select(Availability)
        .where(Availability.room_type_id == room_type.id, Availability.date.in_(nights))
        .with_for_update()
    )
    availability_by_date = {row.date: row for row in lock_result.scalars().all()}

    for day in nights:
        row = availability_by_date.get(day)
        capacity = row.rooms_available if row else 0
        held = await _held_count(session, room_type.id, day)
        if enforce_capacity and capacity - held < 1:
            raise HTTPException(status.HTTP_409_CONFLICT, f"No rooms available on {day.isoformat()}")

    return property_, room_type, rate_plan, nights


async def _upsert_guest(session: AsyncSession, guest_in: GuestIn) -> Guest:
    guest_result = await session.execute(select(Guest).where(Guest.email == guest_in.email))
    guest = guest_result.scalar_one_or_none()
    if guest is None:
        guest = Guest(name=guest_in.name, email=guest_in.email, phone=guest_in.phone)
        if guest_in.id_document_url:
            guest.id_document_url = guest_in.id_document_url
            guest.verification_status = GuestVerificationStatus.pending
        session.add(guest)
        await session.flush()
    elif guest_in.id_document_url and guest_in.id_document_url != guest.id_document_url:
        # A newly submitted document supersedes any prior review — including a
        # past rejection — and needs to be looked at again.
        guest.id_document_url = guest_in.id_document_url
        if guest.verification_status != GuestVerificationStatus.verified:
            guest.verification_status = GuestVerificationStatus.pending
        session.add(guest)
    return guest


async def create_booking(session: AsyncSession, payload: BookingCreate) -> Booking:
    property_, room_type, rate_plan, nights = await _validate_and_lock(
        session,
        property_id=payload.property_id,
        room_type_id=payload.room_type_id,
        rate_plan_id=payload.rate_plan_id,
        check_in=payload.check_in,
        check_out=payload.check_out,
        guests_count=payload.guests_count,
        enforce_capacity=payload.payment_method == BookingSource.online,
    )
    guest = await _upsert_guest(session, payload.guest)

    total_amount = rate_plan.fixed_price * len(nights)
    status_ = (
        BookingStatus.pending_payment
        if payload.payment_method == BookingSource.online
        else BookingStatus.pending_whatsapp
    )

    booking = Booking(
        property_id=property_.id,
        room_type_id=room_type.id,
        rate_plan_id=rate_plan.id,
        guest_id=guest.id,
        check_in=payload.check_in,
        check_out=payload.check_out,
        guests_count=payload.guests_count,
        status=status_,
        source=payload.payment_method,
        total_amount=total_amount,
    )
    session.add(booking)
    await session.flush()

    if payload.payment_method == BookingSource.online:
        session.add(
            Payment(
                booking_id=booking.id,
                gateway=PaymentGateway.stub,
                amount=total_amount,
                status=PaymentStatus.pending,
            )
        )

    await session.commit()
    await session.refresh(booking)
    return booking


async def create_manual_booking(session: AsyncSession, property_id: int, payload: AdminBookingCreate) -> Booking:
    """Staff-entered booking for a walk-in or phone reservation — payment is
    handled outside the system (cash/card at the desk), so it's created
    already confirmed and always counts against real availability."""
    property_, room_type, rate_plan, nights = await _validate_and_lock(
        session,
        property_id=property_id,
        room_type_id=payload.room_type_id,
        rate_plan_id=payload.rate_plan_id,
        check_in=payload.check_in,
        check_out=payload.check_out,
        guests_count=payload.guests_count,
        enforce_capacity=True,
    )
    guest = await _upsert_guest(session, payload.guest)

    booking = Booking(
        property_id=property_.id,
        room_type_id=room_type.id,
        rate_plan_id=rate_plan.id,
        guest_id=guest.id,
        check_in=payload.check_in,
        check_out=payload.check_out,
        guests_count=payload.guests_count,
        status=BookingStatus.confirmed,
        source=BookingSource.manual,
        total_amount=rate_plan.fixed_price * len(nights),
    )
    session.add(booking)
    await session.commit()
    await session.refresh(booking)
    return booking


async def confirm_stub_payment(session: AsyncSession, booking_id: int, email: str) -> Booking:
    """Stub stand-in for the Razorpay webhook. Booking ids are sequential and
    guessable, so this still checks the requester knows the guest's email —
    the same ownership proof GET /api/bookings/{id} requires — otherwise
    anyone could "pay" for anyone else's pending booking."""
    booking = await session.get(Booking, booking_id)
    if not booking:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Booking not found")
    guest = await session.get(Guest, booking.guest_id) if booking.guest_id else None
    if not guest or guest.email.lower() != email.lower():
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Booking not found")
    if booking.status != BookingStatus.pending_payment:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "Booking is not awaiting payment")

    booking.status = BookingStatus.confirmed
    booking.updated_at = datetime.utcnow()
    session.add(booking)

    payment_result = await session.execute(select(Payment).where(Payment.booking_id == booking_id))
    payment = payment_result.scalar_one_or_none()
    if payment:
        payment.status = PaymentStatus.succeeded
        session.add(payment)

    await session.commit()
    await session.refresh(booking)
    return booking


async def cancel_own_booking(session: AsyncSession, booking_id: int, email: str) -> Booking:
    """Guest self-service cancel — same email-ownership proof as confirm_stub_payment
    and GET /api/bookings/{id}, since booking ids are sequential and guessable."""
    booking = await session.get(Booking, booking_id)
    if not booking:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Booking not found")
    guest = await session.get(Guest, booking.guest_id) if booking.guest_id else None
    if not guest or guest.email.lower() != email.lower():
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Booking not found")
    if booking.status not in CANCELLABLE_STATUSES:
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "This booking can no longer be cancelled")
    if booking.check_in <= datetime.utcnow().date():
        raise HTTPException(status.HTTP_400_BAD_REQUEST, "This booking's stay has already started")

    booking.status = BookingStatus.cancelled
    booking.updated_at = datetime.utcnow()
    session.add(booking)
    await session.commit()
    await session.refresh(booking)
    return booking

File: backend/src/services/booking.ts
import { PoolClient } from "pg";
import { settings } from "../config";
import { withTransaction } from "../db";
import { HttpError } from "../middleware/errors";
import { AdminBookingCreate } from "../schemas/adminContent";
import { BookingCreate, GuestIn } from "../schemas/public";
import { BookingSource, BookingStatus, GuestVerificationStatus, PaymentGateway, PaymentStatus } from "../types/enums";
import { Booking, Guest, Property, RatePlan, RoomType } from "../types/models";

// Statuses that count as "holding" a room for a given night.
const HOLDING_STATUSES = [BookingStatus.confirmed, BookingStatus.pending_payment];
// A guest can cancel a booking themselves as long as it's still in one of these
// states and their stay hasn't already started.
const CANCELLABLE_STATUSES = [BookingStatus.pending_payment, BookingStatus.confirmed, BookingStatus.pending_whatsapp];
const MAX_STAY_NIGHTS = 30;
const MAX_ADVANCE_BOOKING_DAYS = 730; // 2 years — keeps the availability loop bounded

function todayDateString(): string {
  return new Date().toISOString().slice(0, 10);
}

function addDays(dateStr: string, days: number): string {
  const d = new Date(dateStr + "T00:00:00Z");
  d.setUTCDate(d.getUTCDate() + days);
  return d.toISOString().slice(0, 10);
}

function daysBetween(a: string, b: string): number {
  const msPerDay = 86_400_000;
  return Math.round((Date.parse(b + "T00:00:00Z") - Date.parse(a + "T00:00:00Z")) / msPerDay);
}

function nightsBetween(checkIn: string, checkOut: string): string[] {
  const count = daysBetween(checkIn, checkOut);
  const nights: string[] = [];
  for (let i = 0; i < count; i++) nights.push(addDays(checkIn, i));
  return nights;
}

async function heldCount(client: PoolClient, roomTypeId: number, day: string): Promise<number> {
  const expiryCutoff = new Date(Date.now() - settings.pendingPaymentExpiryMinutes * 60_000);
  const rows = await client.query<{ count: string }>(
    `SELECT COUNT(*) AS count FROM bookings
     WHERE room_type_id = $1
       AND check_in <= $2
       AND check_out > $2
       AND status = ANY($3::text[])
       AND NOT (status = $4 AND created_at < $5)`,
    [roomTypeId, day, HOLDING_STATUSES, BookingStatus.pending_payment, expiryCutoff]
  );
  return parseInt(rows.rows[0].count, 10);
}

interface ValidatedBookingInputs {
  property: Property;
  roomType: RoomType;
  ratePlan: RatePlan;
  nights: string[];
}

async function validateAndLock(
  client: PoolClient,
  params: {
    propertyId: number;
    roomTypeId: number;
    ratePlanId: number;
    checkIn: string;
    checkOut: string;
    guestsCount: number;
    enforceCapacity: boolean;
  }
): Promise<ValidatedBookingInputs> {
  const { propertyId, roomTypeId, ratePlanId, checkIn, checkOut, guestsCount, enforceCapacity } = params;

  if (checkIn >= checkOut) throw new HttpError(400, "check_out must be after check_in");
  const today = todayDateString();
  if (checkIn < today) throw new HttpError(400, "check_in cannot be in the past");
  if (daysBetween(checkIn, checkOut) > MAX_STAY_NIGHTS) {
    throw new HttpError(400, `Stays longer than ${MAX_STAY_NIGHTS} nights aren't supported`);
  }
  if (daysBetween(today, checkIn) > MAX_ADVANCE_BOOKING_DAYS) {
    throw new HttpError(400, "Check-in date is too far in the future");
  }

  const roomTypeRes = await client.query<RoomType>("SELECT * FROM room_types WHERE id = $1", [roomTypeId]);
  const ratePlanRes = await client.query<RatePlan>("SELECT * FROM rate_plans WHERE id = $1", [ratePlanId]);
  const propertyRes = await client.query<Property>("SELECT * FROM properties WHERE id = $1", [propertyId]);
  const roomType = roomTypeRes.rows[0];
  const ratePlan = ratePlanRes.rows[0];
  const property = propertyRes.rows[0];
  if (!roomType || !ratePlan || !property) {
    throw new HttpError(404, "Property, room type, or rate plan not found");
  }
  if (ratePlan.room_type_id !== roomType.id || roomType.property_id !== property.id) {
    throw new HttpError(400, "Room/rate plan does not belong to this property");
  }
  if (guestsCount > roomType.max_occupancy) {
    throw new HttpError(400, "Guest count exceeds room's max occupancy");
  }
  if (!(ratePlan.valid_from <= checkIn && checkOut <= ratePlan.valid_to)) {
    throw new HttpError(400, "Dates fall outside this rate plan's validity window");
  }

  const nights = nightsBetween(checkIn, checkOut);

  // Lock the relevant availability rows so concurrent bookings for the same
  // room/date range serialize instead of both reading a stale remaining count.
  const lockRes = await client.query<{ date: string; rooms_available: number }>(
    `SELECT date, rooms_available FROM availability
     WHERE room_type_id = $1 AND date = ANY($2::date[])
     FOR UPDATE`,
    [roomType.id, nights]
  );
  const availabilityByDate = new Map(lockRes.rows.map((r) => [r.date, r.rooms_available]));

  for (const day of nights) {
    const capacity = availabilityByDate.get(day) ?? 0;
    const held = await heldCount(client, roomType.id, day);
    if (enforceCapacity && capacity - held < 1) {
      throw new HttpError(409, `No rooms available on ${day}`);
    }
  }

  return { property, roomType, ratePlan, nights };
}

async function upsertGuest(client: PoolClient, guestIn: GuestIn): Promise<Guest> {
  const existingRes = await client.query<Guest>("SELECT * FROM guests WHERE email = $1", [guestIn.email]);
  let guest = existingRes.rows[0];

  if (!guest) {
    const verificationStatus = guestIn.id_document_url ? GuestVerificationStatus.pending : GuestVerificationStatus.unverified;
    const inserted = await client.query<Guest>(
      `INSERT INTO guests (name, email, phone, id_document_url, verification_status)
       VALUES ($1, $2, $3, $4, $5) RETURNING *`,
      [guestIn.name, guestIn.email, guestIn.phone, guestIn.id_document_url ?? null, verificationStatus]
    );
    guest = inserted.rows[0];
  } else if (guestIn.id_document_url && guestIn.id_document_url !== guest.id_document_url) {
    // A newly submitted document supersedes any prior review — including a
    // past rejection — and needs to be looked at again.
    const newStatus =
      guest.verification_status !== GuestVerificationStatus.verified ? GuestVerificationStatus.pending : guest.verification_status;
    const updated = await client.query<Guest>(
      `UPDATE guests SET id_document_url = $1, verification_status = $2 WHERE id = $3 RETURNING *`,
      [guestIn.id_document_url, newStatus, guest.id]
    );
    guest = updated.rows[0];
  }
  return guest;
}

export async function createBooking(payload: BookingCreate): Promise<Booking> {
  return withTransaction(async (client) => {
    const { property, roomType, ratePlan, nights } = await validateAndLock(client, {
      propertyId: payload.property_id,
      roomTypeId: payload.room_type_id,
      ratePlanId: payload.rate_plan_id,
      checkIn: payload.check_in,
      checkOut: payload.check_out,
      guestsCount: payload.guests_count,
      enforceCapacity: payload.payment_method === BookingSource.online,
    });
    const guest = await upsertGuest(client, payload.guest);

    const totalAmount = ratePlan.fixed_price * nights.length;
    const status = payload.payment_method === BookingSource.online ? BookingStatus.pending_payment : BookingStatus.pending_whatsapp;

    const bookingRes = await client.query<Booking>(
      `INSERT INTO bookings (property_id, room_type_id, rate_plan_id, guest_id, check_in, check_out, guests_count, status, source, total_amount)
       VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10) RETURNING *`,
      [
        property.id,
        roomType.id,
        ratePlan.id,
        guest.id,
        payload.check_in,
        payload.check_out,
        payload.guests_count,
        status,
        payload.payment_method,
        totalAmount,
      ]
    );
    const booking = bookingRes.rows[0];

    if (payload.payment_method === BookingSource.online) {
      await client.query(
        `INSERT INTO payments (booking_id, gateway, amount, status) VALUES ($1,$2,$3,$4)`,
        [booking.id, PaymentGateway.stub, totalAmount, PaymentStatus.pending]
      );
    }

    return booking;
  });
}

/** Staff-entered booking for a walk-in or phone reservation — payment is
 * handled outside the system (cash/card at the desk), so it's created
 * already confirmed and always counts against real availability. */
export async function createManualBooking(propertyId: number | null, payload: AdminBookingCreate): Promise<Booking> {
  if (propertyId === null) throw new HttpError(400, "property_id is required");
  return withTransaction(async (client) => {
    const { property, roomType, ratePlan, nights } = await validateAndLock(client, {
      propertyId,
      roomTypeId: payload.room_type_id,
      ratePlanId: payload.rate_plan_id,
      checkIn: payload.check_in,
      checkOut: payload.check_out,
      guestsCount: payload.guests_count,
      enforceCapacity: true,
    });
    const guest = await upsertGuest(client, payload.guest);

    const totalAmount = ratePlan.fixed_price * nights.length;
    const bookingRes = await client.query<Booking>(
      `INSERT INTO bookings (property_id, room_type_id, rate_plan_id, guest_id, check_in, check_out, guests_count, status, source, total_amount)
       VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10) RETURNING *`,
      [
        property.id,
        roomType.id,
        ratePlan.id,
        guest.id,
        payload.check_in,
        payload.check_out,
        payload.guests_count,
        BookingStatus.confirmed,
        BookingSource.manual,
        totalAmount,
      ]
    );
    return bookingRes.rows[0];
  });
}

/** Stub stand-in for the Razorpay webhook. Booking ids are sequential and
 * guessable, so this still checks the requester knows the guest's email —
 * the same ownership proof GET /api/bookings/{id} requires — otherwise
 * anyone could "pay" for anyone else's pending booking. */
export async function confirmStubPayment(bookingId: number, email: string): Promise<Booking> {
  return withTransaction(async (client) => {
    const bookingRes = await client.query<Booking>("SELECT * FROM bookings WHERE id = $1", [bookingId]);
    const booking = bookingRes.rows[0];
    if (!booking) throw new HttpError(404, "Booking not found");
    const guest = booking.guest_id ? (await client.query<Guest>("SELECT * FROM guests WHERE id = $1", [booking.guest_id])).rows[0] : null;
    if (!guest || guest.email.toLowerCase() !== email.toLowerCase()) throw new HttpError(404, "Booking not found");
    if (booking.status !== BookingStatus.pending_payment) throw new HttpError(400, "Booking is not awaiting payment");

    const updated = await client.query<Booking>(
      `UPDATE bookings SET status = $1, updated_at = now() WHERE id = $2 RETURNING *`,
      [BookingStatus.confirmed, bookingId]
    );
    await client.query(`UPDATE payments SET status = $1 WHERE booking_id = $2`, [PaymentStatus.succeeded, bookingId]);
    return updated.rows[0];
  });
}

/** Guest self-service cancel — same email-ownership proof as confirmStubPayment
 * and GET /api/bookings/{id}, since booking ids are sequential and guessable. */
export async function cancelOwnBooking(bookingId: number, email: string): Promise<Booking> {
  return withTransaction(async (client) => {
    const bookingRes = await client.query<Booking>("SELECT * FROM bookings WHERE id = $1", [bookingId]);
    const booking = bookingRes.rows[0];
    if (!booking) throw new HttpError(404, "Booking not found");
    const guest = booking.guest_id ? (await client.query<Guest>("SELECT * FROM guests WHERE id = $1", [booking.guest_id])).rows[0] : null;
    if (!guest || guest.email.toLowerCase() !== email.toLowerCase()) throw new HttpError(404, "Booking not found");
    if (!CANCELLABLE_STATUSES.includes(booking.status)) throw new HttpError(400, "This booking can no longer be cancelled");
    if (booking.check_in <= todayDateString()) throw new HttpError(400, "This booking's stay has already started");

    const updated = await client.query<Booking>(
      `UPDATE bookings SET status = $1, updated_at = now() WHERE id = $2 RETURNING *`,
      [BookingStatus.cancelled, bookingId]
    );
    return updated.rows[0];
  });
}
